source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+            required = False
+            for definition in artifact.defines:
+                if definition in discovery:
+                    new_discovery[definition] = "Seen"
+                    required = True
+
+            # Assuming it is needed, check its
+            # dependencies to know what needs doing
+            if required:
+                compiled = [False]*len(artifact.depends_on)
+                for idep, dependency in enumerate(artifact.depends_on):
+                    if dependency in discovery:
+                        # Are the dependencies compiled?
+                        if discovery[dependency] == "Compiled":
+                            compiled[idep] = True
+                    else:
+                        # If the dependency isn't in the list at all yet
+                        # then add an entry so the system knows we are
+                        # expecting it later (for the above check)
+                        new_discovery[dependency] = "HeardOf"
+
+                # If the dependencies are satisfied (or there weren't
+                # any) then the file can be compiled now
+                if len(compiled) == 0 or all(compiled):
+                    for definition in artifact.defines:
+                        task = self._taskmap[(artifact.filetype,
+                                              artifact.state)]
+                        new_artifacts.extend(task.run([artifact]))
+                        new_discovery[definition] = "Compiled"
+                else:
+                    # If the dependencies weren't all satisfied then
+                    # back on the queue for another pass later
+                    new_artifacts.append(artifact)
+            else:
+                # If it wasn't required it could be later, so
+                # put it back on the queue, unless the target
+                # has been compiled, in which case it wasn't
+                # needed at all!
+                if discovery[self._target] != "Compiled":
+                    new_artifacts.append(artifact)
+
+        elif artifact.state is Compiled:
+            # Begin populating the list for linking
+            new_objects.append(artifact)
+            # But do not return a new artifact - this object
+            # is "done" as far as the processing is concerned
+
+            # But, if this is the file containing the target
+            # that means everything must have been compiled
+            # by this point; so we can do the linking step
+            if self._target in artifact.defines:
+                task = self._taskmap[(artifact.filetype,
+                                      artifact.state)]
+                new_artifacts.extend(task.run(objects + [artifact]))
+
+        elif artifact.state is Linked:
+            # Nothing to do at present with the final linked
+            # executable, but included here for completeness
+            pass
```

Comment:
In the future we might want to run it. e.g. for unit testing.

Reply:
I thought _perhaps_ we might find some other use/s for it later - there could be a variety of things you might want to do as each target is produced (once we have multiple targets)